proto/kuksa/val/v2/val.proto

```diff
@@ -0,0 +1,208 @@
+  //
+  // Returns (GRPC error code):
+  //   NOT_FOUND if any of the signals are non-existant.
+  //   PERMISSION_DENIED if access is denied for any of the signals.
+  rpc PublishValue(PublishValueRequest) returns (PublishValueResponse);
+
+  // Open a stream used to provide actuation and/or publishing values using
+  // a streaming interface. Used to provide actuators and to enable high frequency
+  // updates of values.
+  //
+  // The open stream is used for request / response type communication between the
+  // provider and server (where the initiator of a request can vary).
+  // Errors are communicated as messages in the stream.
+  rpc OpenProviderStream(stream OpenProviderStreamRequest) returns (stream OpenProviderStreamResponse);
+
+  // Get server information
+  rpc GetServerInfo(GetServerInfoRequest) returns (GetServerInfoResponse);
+}
+
+message GetValueRequest {
+  string path = 1;
+}
+
+message GetValueResponse {
+  Datapoint data_point = 1;
+}
+
+message GetValuesRequest {
+  repeated string paths = 1;
+}
+
+message GetValuesResponse {
+  repeated Datapoint datapoints = 1;
+}
+
+message SubscribeRequest {
+  repeated string paths = 1;
+}
+
+message SubscribeResponse {
+  map<string, Datapoint> entries = 1;
+}
+
+message ActuateRequest {
+  string path = 1;
+  Value value = 2;
+}
+
+message ActuateResponse {
+}
+
+message BatchActuateRequest {
+  repeated ActuateRequest actuate_requests = 1;
+}
+
+message BatchActuateResponse {
+}
+
+message PublishValueRequest {
+  // Unique request id that can be used to identify the response.
+  int32 request_id = 1;
+
+  // Signal identified by either id or path
+  oneof signal {
+    int32 id = 10;
+    string path = 11;
+  }
+
+  // The value to publish
+  oneof data {
+    Value value = 20;
+    Failure failure = 21;
+  }
+}
+
+message PublishValueResponse {
+  int32 request_id = 1;
+  // TODO: Add error cases
+}
+
+message OpenProviderStreamRequest {
+  oneof action {
+    // Inform server of a sensor this provider provides.
+    ProvideSensorRequest provide_sensor_request = 1;
+    // Inform server of an actuator this provider provides.
+    ProvideActuatorRequest provide_actuator_request = 2;
+    // Publish a value.
+    PublishValueRequest publish_value_request = 3;
+    // Sent to acknowledge the acceptance of a actuate request.
+    ActuateResponse actuate_response = 4;
+    // Sent to acknowledge the acceptance of a batch actuate
+    // request.
+    BatchActuateResponse batch_actuate_response = 5;
+  }
+}
+
+message OpenProviderStreamResponse {
+  oneof action {
+    // Response to a provide sensor request.
+    ProvideSensorResponse provide_sensor_response = 1;
+    // Response to a provide actuator request.
+    ProvideActuatorResponse provide_actuator_response = 2;
+    // Acknowledgement that a published value was received.
+    PublishValueResponse publish_value_response = 3;
+    // Send an actuate request to a provider.
+    ActuateRequest actuate_request = 4;
```

Comment:
Do we need this single actuator request? Impl-wise, I could imagine, having just the BatchActuate could be simpler